membrane_curvature/lib/mods.py

```diff
@@ -24,66 +38,58 @@ def grid_map(coords, factor):
     return index_grid_l, index_grid_m
 
 
-def core_fast_leaflet(universe, z_Ref, n_cells, selection, max_width):
+def derive_surface(n_cells, selection, max_width):
     """
-    Runs core_fast_leaflet for selected surface
+    Derive surface from distribution of z coordinates in grid.
 
     Parameters
     ----------
-    universe: MDA Universe
-        Provides coordinates and trajectory of the coordinates.
-    z_Ref : dict.
-        Dictionary to store values in every iteration over frames.
     n_cells : int.
         number of cells in the grid of size `max_width`.
     selection : AtomGroup
-        AtomGroup of reference selection to define the surface
+        AtomGroup of reference selected to define the surface
         of the membrane.
     max_width : int.
         Maximum width of simulation box.
 
     Returns
     -------
-
-    Returns a dictionary with tuples of [i,j] as keys, where 0<i<`max_width`,
-    and 0<j<`max_width`. The resulting values for each key are a list of
-    `len=(n_frames)` containing the extracted z_coordinate a given [i,j] cell
-    in each frame.
+    Returns set of z coordinates in grid.
 
     """
+    z_ref = np.zeros([n_cells, n_cells])
+    grid_z_coordinates = np.zeros([n_cells, n_cells])
+    grid_norm_unit = np.zeros([n_cells, n_cells])
 
-    grid_count_frames = np.zeros([n_cells, n_cells])
+    factor = np.float32(n_cells / max_width)
```

Comment:
Is there a special reason for casting to np.float32? Why not just `n_cells / max_width` to keep it simple?

My reason is that you remove something that's mysterious not obvious why it needs to be there so that reading the code becomes easier. If at some point you _need_ an array top be float32, explicitly set the array's dtype.

Reply:
Thanks for the comment @orbeckst Yes, there is a reason to use the `np.float32`. I discovered in the process of writing the code that the `/` operation is more expensive when I perform `n_cells / max_width` due to the type of the variables: `int` and `numpy.float64`, respectively. The `np.float32` was introduced to speed up. Since it is an operation that runs in every frame, the difference in performance is significant. To give you an estimate, in a system with only 500 lipid headgroups to define the surfaces per leaflet, the calculations runs 20% faster using the `np.float32` trick.

Reply:
I think I'm mis-reading this, aren't `max_width` and `n_cells` both `ints` (going by both the docstring and the tests)?

Or are you referring to another division somewhere else?

Reply:
@fiona-naughton just pointed out the following to me: https://github.com/MDAnalysis/membrane-curvature/blob/97713b39675940ea141b2980ae948db74fa3d5e5/membrane_curvature/core.py#L27

In any case, casting to np.float32 is more expensive for this specific operation on my end, does the performance difference maybe appear later on in the calculation?

Reply:
Thanks @IAlibay . In real-world systems the performance is significantly better using `np.float32`. The tests added so far were simplified as suggested by @lilyminium (See #22 and #31). I would prefer to keep it in the code, even if for the tests it looks like overkill. (I can remove it later if the benchmarks say otherwise)

Reply:
I still heavily recommend against early optimization, especially given that we don't seem to know what is actually benefitting from the cast? (it 100% can't be line 64)

Anyways, at the end of the day I don't see it as a big deal (as long as it doesn't impact accuracy), so I'll default to the other mentor's opinions here (especially given that @lilyminium's suggestion https://github.com/MDAnalysis/membrane-curvature/pull/40#discussion_r661054536 would change this code completely?).

However, I will add the following as things that probably need addressing:

1. Both `n_cells` and `max_width` are written in the docstring as `int` inputs. Is this meant to be the case? If not, the docstring needs to be updated accordingly.

2. Especially if they are meant to be `int`, then assuming we still want to keep this optimization in, why `np.float32`? Unless you're doing heavy vector operations, shouldn't a built-in `float` for a single value always be equal or faster in performance? (if they are `int` then it will default to a built-in `float` so it shouldn't even be necessary to cast?)

Reply:
Thanks for highlighting the docstrings, `max_width` is not `int` type, it's `float`. I'll update them. 
I don't see anything wrong with keeping it anyway. I'll leave and will re-evaluate in a performance test.